lib/grandpa/round_test.go

```diff
@@ -216,388 +218,775 @@ func TestGrandpa_DifferentChains(t *testing.T) {
-	for v := range from {
-		for _, tc := range to {
-			if *done {
-				return
-			}
+func TestPlayGrandpaRound(t *testing.T) {
+	ed25519Keyring, err := keystore.NewEd25519Keyring()
+	require.NoError(t, err)
 
-			tc <- &networkVoteMessage{
-				msg: v.(*VoteMessage),
+	tests := map[string]struct {
+		voters          []*ed25519.Keypair
+		whoEquivocates  map[int]struct{}
+		defineBlockTree func(t *testing.T, blockState BlockState, neighbourServices []*Service)
+	}{
+		// this asserts that all validators finalise the same block if they all see the
+		// same pre-votes and pre-commits, even if their chains are different lengths
+		"base_case": {
+			voters: []*ed25519.Keypair{
+				ed25519Keyring.Alice().(*ed25519.Keypair),
+				ed25519Keyring.Bob().(*ed25519.Keypair),
+				ed25519Keyring.Charlie().(*ed25519.Keypair),
+				ed25519Keyring.Ian().(*ed25519.Keypair),
+				ed25519Keyring.George().(*ed25519.Keypair),
+			},
+			defineBlockTree: func(t *testing.T, blockState BlockState, _ []*Service) {
+				const withBranches = false
+				const baseLenght = 4
+				state.AddBlocksToState(t, blockState.(*state.BlockState), baseLenght, withBranches)
+			},
+		},
+
+		"varying_chain": {
+			voters: []*ed25519.Keypair{
+				ed25519Keyring.Alice().(*ed25519.Keypair),
+				ed25519Keyring.Bob().(*ed25519.Keypair),
+				ed25519Keyring.Charlie().(*ed25519.Keypair),
+				ed25519Keyring.Ian().(*ed25519.Keypair),
+			},
+			defineBlockTree: func(t *testing.T, blockState BlockState, neighbourServices []*Service) {
+				const diff uint = 5
+				rand := uint(rand.Intn(int(diff)))
+
+				const withBranches = false
+				const baseLenght = 4
+				headers, _ := state.AddBlocksToState(t, blockState.(*state.BlockState),
+					baseLenght+rand, withBranches)
+
+				// sync the created blocks with the neighbour services
+				// letting them know about those blocks
+				for _, ns := range neighbourServices {
+					for _, header := range headers {
+						block := &types.Block{
+							Header: *header,
+							Body:   types.Body{},
+						}
+						ns.blockState.(*state.BlockState).AddBlock(block)
+					}
+				}
+			},
+		},
+
+		"with_equivocations": {
+			voters: []*ed25519.Keypair{
+				ed25519Keyring.Alice().(*ed25519.Keypair),
+				ed25519Keyring.Bob().(*ed25519.Keypair),
+				ed25519Keyring.Charlie().(*ed25519.Keypair),
+				ed25519Keyring.Dave().(*ed25519.Keypair),
+				ed25519Keyring.Ian().(*ed25519.Keypair),
+			},
+			// alice and charlie equivocates
+			// it is a map as it is easy to check
+			whoEquivocates: map[int]struct{}{
+				3: {},
+				4: {},
+			},
+			defineBlockTree: func(t *testing.T, blockState BlockState, _ []*Service) {
+				// this creates a tree with 2 branches starting at depth 2
+				branches := map[uint]int{2: 1}
+				const baseLenght = 4
+				state.AddBlocksToStateWithFixedBranches(t, blockState.(*state.BlockState), baseLenght, branches)
+			},
+		},
+	}
+
+	for tname, tt := range tests {
+		tt := tt
+		t.Run(tname, func(t *testing.T) {
```

Comment:
nit can we run subtests in parallel? 🤔 

```suggestion
		t.Run(tname, func(t *testing.T) {
			t.Parallel()
```